test/core/QCheck_expect_test.ml

```diff
@@ -127,6 +136,70 @@ module Generator = struct
       ~name:"tree_rev_is_involutive"
       QCheck.(make IntTree.gen_tree)
       (fun tree -> IntTree.(rev_tree (rev_tree tree)) = tree)
+
+  let nat_split2_spec =
+    Test.make ~name:"nat_split2 spec"
+      (make
+        ~print:Print.(pair int (pair int int))
+         Gen.(small_nat >>= fun n ->
+              pair (return n) (nat_split2 n)))
+      (fun (n, (a, b)) ->
+         0 <= a && 0 <= b && a + b = n)
+
+  let pos_split2_spec =
+    Test.make ~name:"pos_split2 spec"
+      (make
+        ~print:Print.(pair int (pair int int))
+         Gen.(small_nat >>= fun n ->
+              (* we need n > 2 *)
+              let n = n + 2 in
+              pair (return n) (pos_split2 n)))
+      (fun (n, (a, b)) ->
+         (0 < a && 0 < b && a + b = n))
+
+  let range_subset_spec =
+    Test.make ~name:"range_subset_spec"
+      (make
+         ~print:Print.(quad int int int (array int))
+         Gen.(pair small_nat small_nat >>= fun (m, n) ->
+              (* we must guarantee [low <= high]
+                 and [size <= high - low + 1] *)
+              let low = m and high = m + n in
+              int_range 0 (high - low + 1) >>= fun size ->
+              quad (return size) (return low) (return high)
+                (range_subset ~size low high)))
+      (fun (size, low, high, arr) ->
+         if size = 0 then arr = [||]
+         else
+           Array.length arr = size
+           && low <= arr.(0)
+           && Array.for_all (fun (a, b) -> a < b)
+               (Array.init (size - 1) (fun k -> arr.(k), arr.(k+1)))
+           && arr.(size - 1) <= high)
+
+  let nat_split =
+    Test.make ~name:"nat_split"
+      (make
+         ~print:Print.(pair int (array int))
+         Gen.(small_nat >>= fun n ->
+              pair (return n) (nat_split ~size:n n)))
```

Comment:
#198 reminded me of this PR - so I was just looking at the code.

If I understand correctly, this tests an `n`-way split of `n`, which is an important case to test! 
To also test the case `size < n` can some `min/max` magic like for `pos_split` (below) be applied?

Reply:
This is a good suggestion, I'll try to apply it shortly, thanks!

Reply:
Note: for `pos_split`, testing the n-way split only would be fairly bad, because it is a very special case where the output must be [1, 1, 1, ...]. For `nat_split`, where some numbers returned can be 0, the n-way split can return arbitrary values in [0; n] in each element, so the n-way split is more representative of any other split.

Reply:
I made a shot at this in 71675b5276372310508885986d5c41bafddce094